### ISSUE TITLE
TLS error (Safari) or warning (FF) with KMS 6.13

### ISSUE THREAD

**Issue by RaistGH** [Kurento/bugtracker#458]
## Issue description
When connecting to the WebSocketSecure:
- No problems in Chrome
- Firefox 75 outputs a warn: `This site uses a deprecated version of TLS that will be disabled in March 2020. Please upgrade to TLS 1.2 or 1.3.`. The system works with the warning by the moment.
- Safari 13.1 outputs an error: `WebSocket network error: OSStatus Error -9801: Cipher Suite negotiation failure`. This error prevents the connection so the system cannot work.

## Context
This happens when trying to establish a WSS connection. No changes in code, only browsers were updated.

## How to reproduce?
Other people don't seem to have the issue:
https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!topic/kurento/WGor53Yg7g8
I need help to track the issue.

1. Brand new KMS 6.13 installation
2. Enable wss and install certificate
3. Use last kurento-client.js (tested also with previous ones)
4. Call constructor "kurentoClient" exported by kurento-client.js with a wss connection

Very basic.

## Expected & current behavior
No warnings or errors should appear, connecting with TLS over 1.0
Safari 13.1 should connect, since it doesn't do it now.

## Environment info
Tested with the following:
- KMS 6.13.0 (Docker)
- KMS 6.13.0 (Ubuntu 16.04)
- KMS 6.13.1 (Ubuntu 16.04)
- kurento-client.js (6.13)
- kurento-client.js (6.12)
- kurento-client.js (6.8)

- No problems in Chrome
- Firefox 75 (Ubuntu 16.04 and Windows 7) outputs a warn: `This site uses a deprecated version of TLS that will be disabled in March 2020. Please upgrade to TLS 1.2 or 1.3.`
- Safari 13.1 outputs an error: `WebSocket network error: OSStatus Error -9801: Cipher Suite negotiation failure`

- Tested using TURN server (Docker KMS 6.13.0 and kurento-client.js 6.13) using domain SSL certificate
- Tested using STUN server in localhost (Ubuntu 16.04 KMS, all kurento-client.js) with self-signed SSL certificate

### INFO: Kurento Media Server
* Kurento version: 6.13.0
* Server OS:  Debian 9.8 as host
* Installation method: Docker

* Kurento version: 6.13.0 & 6.13.1
* Server OS:  Xenial in VirtualMachine with Windows 7 as host
* Installation method: apt-get

### INFO: Application Server

* Language: browser JavaScript
* Kurento Client version: 6.12.0, 6.13.0, 6.8.0

### INFO: End-user clients

* Device(s):    PC, Mac, iPhone
* OS(es):       Ubuntu 16.04, Windows 10, Windows 7, iOS 13
* Browser(s):  Firefox 75, Chrome 80, Safari 13.1

### Version details
Docker
```
**lsb_release -a (Host)**
No LSB modules are available.
Distributor ID:	Debian
Description:	Debian GNU/Linux 9.8 (stretch)
Release:	9.8
Codename:	stretch
```
```
**kurento-media-server -v (Container)**
0:00:00.047664150 15760 0x5630e409b8a0 INFO    KurentoModuleManager ModuleManager.cpp:90:loadModule: Load file: /usr/lib/x86_64-linux-gnu/kurento/modules/libkmscoremodule.so, module name: /usr/lib/x86_64-linux-gnu/kurento/modules/libkmscoremodule.so
0:00:00.051334232 15760 0x5630e409b8a0 INFO    KurentoModuleManager ModuleManager.cpp:137:loadModule: Loaded module: core, version: 6.13.0, date: Dec 17 2019 13:17:31
0:00:00.052573436 15760 0x5630e409b8a0 INFO    KurentoModuleManager ModuleManager.cpp:90:loadModule: Load file: /usr/lib/x86_64-linux-gnu/kurento/modules/libkmsfiltersmodule.so, module name: /usr/lib/x86_64-linux-gnu/kurento/modules/libkmsfiltersmodule.so
0:00:00.052613534 15760 0x5630e409b8a0 INFO    KurentoModuleManager ModuleManager.cpp:137:loadModule: Loaded module: filters, version: 6.13.0, date: Dec 17 2019 13:35:00
0:00:00.070030806 15760 0x5630e409b8a0 INFO    KurentoModuleManager ModuleManager.cpp:90:loadModule: Load file: /usr/lib/x86_64-linux-gnu/kurento/modules/libkmselementsmodule.so, module name: /usr/lib/x86_64-linux-gnu/kurento/modules/libkmselementsmodule.so
0:00:00.070093135 15760 0x5630e409b8a0 INFO    KurentoModuleManager ModuleManager.cpp:137:loadModule: Loaded module: elements, version: 6.13.0, date: Dec 17 2019 13:26:02
Kurento Media Server version: 6.13.0
Found modules:
	'core' version 6.13.0
	'elements' version 6.13.0
	'filters' version 6.13.0
```
```
**dpkg -l | grep -Pi 'kurento|kms-|gst.*1.5|nice' (Container)**
ii  gstreamer1.5-libav:amd64             1.8.1-1kurento1.16.04                    amd64        libav plugin for GStreamer
ii  gstreamer1.5-nice:amd64              0.1.15-1kurento3.16.04                   amd64        ICE library (GStreamer 1.5 plugin)
ii  gstreamer1.5-plugins-bad:amd64       1.8.1-1kurento4.16.04                    amd64        GStreamer plugins from the "bad" set
ii  gstreamer1.5-plugins-base:amd64      1.8.1-1kurento2.16.04                    amd64        GStreamer plugins from the "base" set
ii  gstreamer1.5-plugins-good:amd64      1.8.1-1kurento3.16.04                    amd64        GStreamer plugins from the "good" set
ii  gstreamer1.5-plugins-ugly:amd64      1.8.1-1kurento1.16.04                    amd64        GStreamer plugins from the "ugly" set
ii  gstreamer1.5-pulseaudio:amd64        1.8.1-1kurento3.16.04                    amd64        GStreamer plugin for PulseAudio
ii  kms-core                             6.13.0-0kurento1.16.04                   amd64        Kurento Core module
ii  kms-elements                         6.13.0-0kurento1.16.04                   amd64        Kurento Elements module
ii  kms-filters                          6.13.0-0kurento1.16.04                   amd64        Kurento Filters module
ii  kms-jsonrpc                          6.13.0-0kurento1.16.04                   amd64        Kurento JSON-RPC library
ii  kmsjsoncpp                           1.6.3-1kurento1.16.04                    amd64        Kurento jsoncpp library
ii  kurento-media-server                 6.13.0-0kurento1.16.04                   amd64        Kurento Media Server
ii  libgstreamer-plugins-bad1.5-0:amd64  1.8.1-1kurento4.16.04                    amd64        GStreamer development files for libraries from the "bad" set
ii  libgstreamer-plugins-base1.5-0:amd64 1.8.1-1kurento2.16.04                    amd64        GStreamer libraries from the "base" set
ii  libgstreamer1.5-0:amd64              1.8.1-1kurento2.16.04                    amd64        Core GStreamer libraries and elements
ii  libnice10:amd64                      0.1.15-1kurento3.16.04                   amd64        ICE library (shared library)
ii  libsrtp0:amd64                       1.6.0-0kurento1.16.04                    amd64        Secure RTP (SRTP) and UST Reference Implementations - shared library
ii  libusrsctp                           0.9.2-1kurento1.16.04                    amd64        sctp-refimpl library
ii  openh264                             1.4.0-1kurento1.16.04                    amd64        OpenH264 library
ii  openh264-gst-plugins-bad-1.5:amd64   1.8.1-1kurento4.16.04                    amd64        GStreamer plugins from openh264
ii  openwebrtc-gst-plugins               0.10.0-1kurento1.16.04                   amd64        OpenWebRTC specific GStreamer plugins
```
Xenial apt-get
```
**lsb_release -a**
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 16.04.6 LTS
Release:	16.04
Codename:	xenial
```
```
**kurento-media-server -v**
Kurento Media Server version: 6.13.0
Found modules:
	'core' version 6.13.0
	'elements' version 6.13.0
	'filters' version 6.13.0
```
```
**dpkg -l | grep -Pi 'kurento|kms-|gst.*1.5|nice'**
ii  gstreamer1.5-libav:amd64                   1.8.1-1kurento1.16.04                                    amd64        libav plugin for GStreamer
ii  gstreamer1.5-nice:amd64                    0.1.15-1kurento3.16.04                                   amd64        ICE library (GStreamer 1.5 plugin)
ii  gstreamer1.5-plugins-bad:amd64             1.8.1-1kurento4.16.04                                    amd64        GStreamer plugins from the "bad" set
ii  gstreamer1.5-plugins-base:amd64            1.8.1-1kurento2.16.04                                    amd64        GStreamer plugins from the "base" set
ii  gstreamer1.5-plugins-good:amd64            1.8.1-1kurento3.16.04                                    amd64        GStreamer plugins from the "good" set
ii  gstreamer1.5-plugins-ugly:amd64            1.8.1-1kurento1.16.04                                    amd64        GStreamer plugins from the "ugly" set
ii  gstreamer1.5-pulseaudio:amd64              1.8.1-1kurento3.16.04                                    amd64        GStreamer plugin for PulseAudio
ii  gstreamer1.5-x:amd64                       1.8.1-1kurento2.16.04                                    amd64        GStreamer plugins for X11 and Pango
ii  kms-core                                   6.13.0-0kurento1.16.04                                   amd64        Kurento Core module
ii  kms-elements                               6.13.0-0kurento1.16.04                                   amd64        Kurento Elements module
ii  kms-filters                                6.13.0-0kurento1.16.04                                   amd64        Kurento Filters module
ii  kms-jsonrpc                                6.13.0-0kurento1.16.04                                   amd64        Kurento JSON-RPC library
ii  kmsjsoncpp                                 1.6.3-1kurento1.16.04                                    amd64        Kurento jsoncpp library
ii  kurento-media-server                       6.13.0-0kurento1.16.04                                   amd64        Kurento Media Server
ii  libgstreamer-plugins-bad1.5-0:amd64        1.8.1-1kurento4.16.04                                    amd64        GStreamer development files for libraries from the "bad" set
ii  libgstreamer-plugins-base1.5-0:amd64       1.8.1-1kurento2.16.04                                    amd64        GStreamer libraries from the "base" set
ii  libgstreamer1.5-0:amd64                    1.8.1-1kurento2.16.04                                    amd64        Core GStreamer libraries and elements
ii  libnice10:amd64                            0.1.15-1kurento3.16.04                                   amd64        ICE library (shared library)
ii  libreoffice-avmedia-backend-gstreamer      1:5.1.6~rc2-0ubuntu1~xenial10                            amd64        GStreamer backend for LibreOffice
ii  libsrtp0:amd64                             1.6.0-0kurento1.16.04                                    amd64        Secure RTP (SRTP) and UST Reference Implementations - shared library
ii  libusrsctp                                 0.9.2-1kurento1.16.04                                    amd64        sctp-refimpl library
ii  openh264                                   1.4.0-1kurento1.16.04                                    amd64        OpenH264 library
ii  openh264-gst-plugins-bad-1.5:amd64         1.8.1-1kurento4.16.04                                    amd64        GStreamer plugins from openh264
ii  openwebrtc-gst-plugins                     0.10.0-1kurento1.16.04                                   amd64        OpenWebRTC specific GStreamer plugins
```

**Comment by RaistGH** [comment]
In order to try to give more information, is there any other element that can be playing with the WSS with TLS?
I've installed a fresh KMS 6.13 to test it and the error is in the constructor of kurento-client.js, so I don't see where anything of my code or environment can be interfering. But obviously it seems like other people don't have the issue:
https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!topic/kurento/WGor53Yg7g8

**Comment by micaelgallego** [comment]
It is recommended to manage KMS from backend side of the app. 

If you connect to the KMS from frontend code, then any user of your app can know how to connect to your KMS instance and make "bad things".

**Comment by RaistGH** [comment]
> It is recommended to manage KMS from backend side of the app.
> 
> If you connect to the KMS from frontend code, then any user of your app can know how to connect to your KMS instance and make "bad things".

Do you mean using NodeJS server?
If so, at the moment isn't possible, although indeed it's one of the things to change asap.
But what would it change for the TLS error?
Thanks

**Comment by micaelgallego** [comment]
Yes.. I mean a backend server implemented with NodeJS or with any other technology.

If you connect to KMS from server instead of brower, I think TLS warning will be gone.

**Comment by RaistGH** [comment]
Mmmm, I agree about the security issue, but:
- The warnings disappear because an "intermediary" server is in place. That only translates the TLS issue to the intermediary. That's not an issue if KMS and the intermediary are the same server, but if they are in different servers (as it would be in our case), it could be. Besides the intermediary has to deal with the security issue. I think we can workaround this whole point tho.
- As per what I understand, [browser Javascript tutorial](https://doc-kurento.readthedocs.io/en/stable/user/tutorials.html) wouldn't be complaint anymore. At least for Firefox and Safari, since they use KurentoClient connecting directly to KMS.
- "other technology" would be only Java EE, since the languages to interact with KMS are JS or Java.

Am I right?

In any case I'd love some hint to create that intermediary/bridge server-side part. I've read [your document about writing Kurento Applications](https://doc-kurento.readthedocs.io/en/stable/user/writing_applications.html) (I think it's new, I don't remember it last year), but I'm a bit lost at this point. Is there any example-code that I can check to fully understand how to work with it?

Thanks a lot.

**Comment by micaelgallego** [comment]
Please find my comments below.

On Sun, Apr 19, 2020 at 8:43 PM RaistGH <notifications@github.com> wrote:

> Mmmm, I agree about the security issue, but:
>
>    - The warnings disappear because an "intermediary" server is in place.
>    That only translates the TLS issue to the intermediary. That's not an issue
>    if KMS and the intermediary are the same server, but if they are in
>    different servers (as it would be in our case), it could be.
>
> It solves the issue because the intermediary likely don't have issues with
TLS connection from browsers (because use modern TLS version). Neither with
KMS, because browsers are complaining about TLS, but your intermediary
likely not.

>
>    - Besides the intermediary has to deal with the security issue. I
>    think we can workaround this whole point tho.
>    - As per what I understand, browser Javascript tutorial
>    <https://doc-kurento.readthedocs.io/en/stable/user/tutorials.html>
>    wouldn't be complaint anymore. At least for Firefox and Safari, since they
>    use KurentoClient connecting directly to KMS.
>
> You are right. Or KMS updates his WebSocket server or this issue will
prevent tutorials working anymore. If fact, these tutorials should be
removed entirely because they are insecure.

>
>    - "other technology" would be only Java EE, since the languages to
>    interact with KMS are JS or Java.
>
> You can use Java or you can implement a KurentoClient yourself
implementing the JSON-RPC over WebSocket protocol implemented by KMS:
https://doc-kurento.readthedocs.io/en/6.13.0/features/kurento_protocol.html

>
>
> Am I right?
>
More or less.

> In any case I'd love some hint to create that intermediary/bridge
> server-side part. I've read your document about writing Kurento
> Applications
> <https://doc-kurento.readthedocs.io/en/stable/user/writing_applications.html>
> (I think it's new, I don't remember it last year), but I'm a bit lost at
> this point. Is there any example-code that I can check to fully understand
> how to work with it?
>
The document has 4 or 5 years ;) Yep, you can use this example:
https://doc-kurento.readthedocs.io/en/stable/tutorials/java/tutorial-helloworld.html
or
any other with a backend.

But I encourage you to use OpenVidu. It is easier to work with.

> Thanks a lot.
>
You are welcome.

**Comment by j1elo** [comment]
[This line in kurento-media-server](https://github.com/Kurento/kurento-media-server/blob/93fee4aee599dd0c2f08265d0adebdb22aa30e3c/server/transport/websocket/WebSocketTransport.cpp#L206) initializes the secure Kurento WebSocket server, by using the websocketpp library, and setting the SSL context `tlsv1`.

In the [Boost library context docs](https://www.boost.org/doc/libs/1_58_0/doc/html/boost_asio/reference/ssl__context/method.html) I can see that there are multiple other possible SSL context settings, and this one seems interesting:

> tlsv12
>    Generic TLS version 1.2.

Maybe changing `tlsv1` to `tlsv12` is enough to solve the issue, thanks to all the underlying libraries doing the actual hard work. If someone can check out the sources to make that change and test, it would be useful knowing if it works.

**Comment by RaistGH** [comment]
> [This line in kurento-media-server](https://github.com/Kurento/kurento-media-server/blob/93fee4aee599dd0c2f08265d0adebdb22aa30e3c/server/transport/websocket/WebSocketTransport.cpp#L206) initializes the secure Kurento WebSocket server, by using the websocketpp library, and setting the SSL context `tlsv1`.
> 
> In the [Boost library context docs](https://www.boost.org/doc/libs/1_58_0/doc/html/boost_asio/reference/ssl__context/method.html) I can see that there are multiple other possible SSL context settings, and this one seems interesting:
> 
> > tlsv12
> > Generic TLS version 1.2.
> 
> Maybe changing `tlsv1` to `tlsv12` is enough to solve the issue, thanks to all the underlying libraries doing the actual hard work. If someone can check out the sources to make that change and test, it would be useful knowing if it works.

Unfortunately it doesn't work, the websocket simply doesn't connect, although I cannot debug why. I've tried also with `tlsv11` and I've also update `websocketpp` to the last version.
I've tried to upgrade to `libboost1.65` in Xenial (which has only until version 1.58) since it is including a generic `tls`, but I couldn't. Maybe I can try from a Bionic dist, but if `tlsv12` didn't work, I don't think `tls` will work either.

> But I encourage you to use OpenVidu. It is easier to work with.
OpenVidu doesn't seem to fit in our project whose main purpose is to record video (in different directories). OpenVidu is more focused on video-conferences and the recordings seem to be less flexible (for example, only one directory).

**Comment by j1elo** [comment]
I was able to confirm the warning message on Firefox 75.0 and Safari 13.1; in summary, this seems to be the browser situation:

* **Firefox 74.0** warns about imminent removal of support for TLS 1.0
* **Chrome 81.0** accepts TLS 1.0 without failure, because [removal of TLS 1.1 has been delayed until Chrome 84](https://www.chromestatus.com/feature/5759116003770368), due to the ongoing virus outbreak, to avoid disrupting WebRTC communications from still unpatched services.
* **Safari 13.0** accepts TLS 1.0 without issues or warnings.
* **Safari 13.1** moved on to reject TLS 1.0.

Note that, unlike with HTTPS, *browsers don't allow at all using untrusted self-signed certs for Secure WebSocket connections* (`wss://`). Using a self-signed cert will cause an `ERR_CERT_AUTHORITY_INVALID` error in Chrome, and silent errors in Firefox and Safari. You must use a properly issued certificate, or install your own self-signed Root CA into the browser trusted store -- e.g. with a tool such as [mkcert](https://github.com/FiloSottile/mkcert).

I have tested using `tlsv12` in the secure websocket server code in Kurento, and after a rebuild it works fine in all mentioned browser versions.

**Comment by j1elo** [comment]
@micaelgallego 
> It solves the issue because the intermediary likely don't have issues with
> TLS connection from browsers (because use modern TLS version). Neither with
> KMS, because browsers are complaining about TLS, but your intermediary
> likely not.

Note that fixing this was still important, because TLS 1.0 is considered insecure, so the link between KMS and an intermediary (the server application) would be vulnerable to attacks.

Also it probably was just a matter of time that Node itself started removing support for TLS 1.0 too...

**Comment by micaelgallego** [comment]
Good job @j1elo

**Comment by RaistGH** [comment]
Wow, awesome, thanks a lot @j1elo !!
I wonder why it was not working on my build... I was using exactly the same change O.o
It was the first time I compiled it, so maybe I was doing something wrong, but just "tlsv1" was working. Never mind. Thanks!

I''ve tested the nightly one and it's working for us too. In what release is going to be merged?

Few considerations for the future:
- AFAIK tlsv12 forces to the browser to use only TLS v1.2, so if any browser supports only v1.3 at some point, the socket won't connect. Old browsers won't work either. They're very old tho:
  - Chrome < v30
  - Firefox < v27
  - IE < v11
  - Safari < v7
  - Microsoft Edge < v12property.
  - Opera < v17
- To solve the situation at some point it's probably needed to use the generic "tls" string, however that's not available with libboost-dev < v1.65, while Xenial only supports v1.58 for development. Therefore builds would need minimum Bionic to be compiled for both local source and docker.

**Comment by j1elo** [comment]
The enum values present in `boost::asio::ssl::context` are a direct match with the corresponding context creation functions from OpenSSL:

* https://www.boost.org/doc/libs/1_58_0/boost/asio/ssl/context_base.hpp
* https://www.openssl.org/docs/man1.0.2/man3/TLSv1_method.html

Here I see that all `TLSv1_*` functions are specific to one version of TLS. So, `tlsv1`, `tlsv11`, and `tlsv12` correspond uniquely to TLS 1.0, 1.1, and 1.2, respectively.

However there is this OpenSSL funcion: `SSLv23_method()`, that says:

> general-purpose version-flexible SSL/TLS methods. The actual protocol version used will be negotiated to the highest version mutually supported by the client and the server. The supported protocols are SSLv2, SSLv3, TLSv1, TLSv1.1 and TLSv1.2. Most applications should use these method, and avoid the version specific methods described below.

Maybe we should just move on to use these *version-flexible* functions? It would mean using this in our code: `boost::asio::ssl::context::sslv23`.

That way, the most recent protocol version will be negotiated between the KMS server and whatever client is trying to connect.

And, when we perform a definitive migration to Ubuntu 18.04 (or now, maybe straight up to 20.04), we can change it to `boost::asio::ssl::context::tls`.

**Comment by RaistGH** [comment]
The version-flexible function seems interesting :) Although it may introduce a delay when opening the websocket. I assume it won't be important tho.

P.S.: Do you have an approximate date for release 6.13.3 including the TLS fix? Thanks!

**Comment by j1elo** [comment]
@RaistGH the change that forces TLS 1.2 is now released with the patch version 6.13.2, you can use it with apt-get repos or Docker images.

After that one I've also tested with the *version-flexible* specification, and it works well, so I just pushed to master a new change that allows KMS to negotiate the highest TLS version that is supported by the client, among TLS 1.0, 1.1, and 1.2. This is in master and won't be published until the next release, possibly Kurento 6.14. However I believe it doesn't affect you as the previous change of forcing TLS 1.2 is good enough for your case.

Probably most clients now already support TLS 1.2, but still, I thought this change is good because it prevents breaking old clients that might still be connecting with older versions of TLS.

Commit: https://github.com/Kurento/kurento-media-server/commit/8f91bb750b3c333908b46682012dd4cb3abfb4b2
